File: ConstructBinaryTreeFromPreorderAndInorderTraversal.py
# Definition for a binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param preorder, a list of integers
    # @param inorder, a list of integers
    # @return a tree node
    def buildTree(self, preorder, inorder):
        if len(preorder) == 0 or len(inorder) == 0:
            return None
        root = TreeNode(preorder[0])
        llen = inorder.index(root.val)  # index to len need +1, but get rid of root, -1
        rlen = len(inorder) - 1 - llen # get rid root, need -1
        ltree = self.buildTree(preorder[1:1+llen], inorder[:llen])
        rtree = self.buildTree(preorder[1+llen:], inorder[llen+1:])
        root.left = ltree
        root.right = rtree
        return root

def preorder(root):
    if root == None:
        return
    print root.val
    preorder(root.left)
    preorder(root.right)

    
if __name__ == '__main__':
    pre = [4,2,1,3,5,6,7]
    ino = [1,2,3,4,5,7,6]
    so = Solution()
    tree = so.buildTree(pre, ino)
    preorder(tree)

File: CPP/Second/construct_binary_tree_from_preorder_and_inorder_traversal.cc
// https://oj.leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/

// Given preorder and inorder traversal of a tree, construct the binary tree.
// Note:
// You may assume that duplicates do not exist in the tree.

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

/*
    Time O(n), Space O(logN)
*/

class Solution {
public:
    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder) {
        return build(&preorder[0], &*preorder.end(), &inorder[0], &*inorder.end());
    }
    
    TreeNode* build(int *pb, int *pe, int *ib, int *ie) {
        if (pb == pe)
            return nullptr;
        auto root = new TreeNode(*pb);
        int left_len = find(ib, ie, root->val) - ib;
        root->left = build(pb+1, pb+1+left_len, ib, ib+left_len);
        root->right = build(pb+1+left_len, pe, ib+left_len+1, ie);
        return root;
    }
};


int main(int argc, char *argv[])
{

    return 0;
}
